Lib/ufoLib/filesystem.py

```diff
@@ -0,0 +1,523 @@
+		if self.isDirectory(path1):
+			self._fs.movedir(path1, path2, create=True)
+		else:
+			self._fs.move(path1, path2)
+
+	def _fsExists(self, path):
+		path = self._fsRootPath(path)
+		return self._fs.exists(path)
+
+	def _fsIsDirectory(self, path):
+		path = self._fsRootPath(path)
+		return self._fs.isdir(path)
+
+	def _fsListDirectory(self, path):
+		path = self._fsRootPath(path)
+		return self._fs.listdir(path)
+
+	def _fsGetFileModificationTime(self, path):
+		path = self._fsRootPath(path)
+		info = self._fs.getinfo(path, namespaces=["details"])
+		return info.modified
+
+	# -----------------
+	# Path Manipulation
+	# -----------------
+
+	def joinPath(self, *parts):
+		if haveFS:
+			return fs.path.join(*parts)
+		else:
+			return os.path.join(*parts)
+
+	def splitPath(self, path):
+		if haveFS:
+			return fs.path.split(path)
+		else:
+			return os.path.split(path)
+
+	def directoryName(self, path):
+		if haveFS:
+			return fs.path.dirname(path)
+		else:
+			return os.path.dirname(path)
+
+	def relativePath(self, path, start):
+		if haveFS:
+			return fs.relativefrom(path, start)
+		else:
+			return os.path.relpath(path, start)
+
+	# ---------
+	# Existence
+	# ---------
+
+	def exists(self, path):
+		return self._fsExists(path)
+
+	def isDirectory(self, path):
+		return self._fsIsDirectory(path)
+
+	def listDirectory(self, path, recurse=False):
+		return self._listDirectory(path, recurse=recurse, relativeTo=path)
+
+	def _listDirectory(self, path, recurse=False, relativeTo=None, depth=0, maxDepth=100):
+		sep = os.sep
+		if not relativeTo.endswith(sep):
+			relativeTo += sep
+		if depth > maxDepth:
+			raise UFOLibError("Maximum recusion depth reached.")
+		result = []
+		for fileName in self._fsListDirectory(path):
+			p = self.joinPath(path, fileName)
+			if self.isDirectory(p) and recurse:
+				result += self._listDirectory(p, recurse=True, relativeTo=relativeTo, depth=depth+1, maxDepth=maxDepth)
+			else:
+				p = p[len(relativeTo):]
+				if sep != "/":
+					# replace '\\' with '/'
+					p = p.replace(sep, "/")
+				result.append(p)
+		return result
+
+	def makeDirectory(self, path):
+		if not self.exists(path):
+			self._fsMakeDirectory(path)
+
+	# -----------
+	# File Opener
+	# -----------
+
+	def open(self, path, mode="r", encoding=None):
+		"""
+		Returns a file (or file-like) object for the
+		file at the given path. The path must be relative
+		to the UFO path. Returns None if the file does
+		not exist and the mode is "r" or "rb. An encoding
+		may be passed if needed.
+
+		Note: The caller is responsible for closing the open file.
+		"""
+		if encoding:
+			if mode == "r":
+				mode = "rb"
+			elif mode == "w":
+				mode = "wb"
+		if mode in ("r", "rb") and not self.exists(path):
+			return None
+		if self.exists(path) and self.isDirectory(path):
+			raise UFOLibError("%s is a directory." % path)
+		if mode in ("w", "wb"):
+			self._buildDirectoryTree(path)
+		f = self._fsOpen(path, mode, encoding=encoding)
```

Comment:
this `FileSystem.open` method is doing too many things, and some of them are just wrong.
first of all, calling the built-in `open` (or `io.open`) with a non-None `encoding` only makes sense for text, not for binary data. `encoding` argument should never be passed when reading/writing binary files. The meaning of `encoding` argument is: when reading a text file, the bytes are decoded to a unicode string using the specified `encoding`; when writing a text file, the unicode string is encoded to bytes using the specified `encoding`. If we say, `mode=rb`, then we get bytes back, no decoding occurs; same for `wb`; we are expected to pass bytes, which will be written as is to file.

Moreover, all this checking if some path exists may be costly and unneeded in some circumstances, i.e. when we know that the file should be there. That's what exceptions are for. An IOError (with errno=2 as in "file does not exist") should be raised in the underlying filesystem's `open` method and if we like we can re-raise it wrapped in some UFOLibError subclass that means (`UFOLibFileNotFoundError`?) and let the caller deal with it as needed. 
And returning `None` when the file doesn't exist is not very Pythonic. Reminds me of NULL and C fopen.

It's funny that every time this very cautios `open` method is used elsewhere, there is a corresponding `self.fileSystem.exists(path)` just before it (probably to avoid an `AttributeError: None has no attribute...` later when attempting to use the context manager value in the with `self.fileSystem.open(...) as fp` block. And then, _again_, we check, twice!, if the path exists (upong reading), or if it's a directory... Why?

Let's clean this up.